### PR TITLE
add means to handle different kubeadm APIs

### PR DESCRIPTION
**Pull request by neolit123**
add handling of the various kubeadm APIs:
- v1alpha1 and v1alpha2 are the same for the current usage
- v1alpha3 is handled slightly differenetly

fixes https://github.com/kubernetes-sigs/kind/issues/7
/kind feature
/assign @BenTheElder

### CODE REVIEW

File: pkg/cluster/kubeadm/config.go [hunk 1]
```diff
@@ -51,25 +52,33 @@ func (c *ConfigData) Derive() {
...
+unifiedControlPlaneImage: {{.UnifiedControlPlaneImage}}:{{.DockerStableTag}}
+{{end}}`
+
+// DefaultConfigTemplateAlphaV3 is the kubadm config temaplte for
+// API version v1alpha3
+const DefaultConfigTemplateAlphaV3 = `# config generated by kind
+apiVersion: kubeadm.k8s.io/v1alpha3
+kind: ClusterConfiguration
 kubernetesVersion: {{.KubernetesVersion}}
-# TODO(bentheelder): fix this upstream!
```
Comment (BenTheElder) [inline pkg/cluster/kubeadm/config.go]: we still need this :(

File: pkg/cluster/node.go [hunk 2]
```diff
@@ -169,7 +169,7 @@ func (nh *nodeHandle) CombinedOutputLines(command string, args ...string) ([]str
 func (nh *nodeHandle) CopyTo(source, dest string) error {
 	cmd := exec.Command("docker", "cp")
 	cmd.Args = append(cmd.Args,
-		source,               // from the source file
+		source, // from the source file
```
[the remark this answers is not in the document]
Comment (BenTheElder) [reply pkg/cluster/node.go]: I mean if not now, we'll need it soon. I imagine k/k should switch to go 1.11 any day now that the release is done.

File: pkg/cluster/kubeadm/config.go [hunk 3]
```diff
@@ -51,25 +52,33 @@ func (c *ConfigData) Derive() {
...
 # on docker for mac we have to expose the api server via port forward,
 # so we need to ensure the cert is valid for localhost so we can talk
 # to the cluster after rewriting the kubeconfig to point to localhost
 apiServerCertSANs: [localhost]
+{{if ne .UnifiedControlPlaneImage ""}}
+# optionally specify a unified control plane image
+unifiedControlPlaneImage: {{.UnifiedControlPlaneImage}}:{{.DockerStableTag}}
+{{end}}`
+
+// DefaultConfigTemplateAlphaV3 is the kubadm config temaplte for
```
Comment (neolit123) [inline pkg/cluster/kubeadm/config.go]: found typo.

File: pkg/cluster/kubeadm/config.go [hunk 4]
```diff
@@ -51,25 +52,33 @@ func (c *ConfigData) Derive() {
 	}
 }
 
-// DefaultConfigTemplate is the default kubeadm config template used by kind
-const DefaultConfigTemplate = `# config generated by kind
+// DefaultConfigTemplateAlphaV1orV2 is the kubadm config temaplte for
```
Comment (neolit123) [inline pkg/cluster/kubeadm/config.go]: typo.

File: pkg/cluster/kubeadm/config.go (hunk 1 above)
Comment (neolit123) [reply pkg/cluster/kubeadm/config.go]: should have left it, my bad.

**Comment by BenTheElder** [comment]
To add a library I think you need `hack/update-deps.sh add some-go/import/path`, sorry I haven't gotten guides yet. I meant to do that by the end of this week but got sidetracked with the DNS breakage.

**Comment by neolit123** [comment]
@BenTheElder updated.
excluded the gofmt bit as i don't want to update to go 1.11 yet.

**Comment by BenTheElder** [comment]
can you use something like [GVM](https://github.com/moovweb/gvm) ? I usually install multiple go versions.

It looks like you're passing verify now anyhow though

**Comment by neolit123** [comment]
is there are good reason for the 1.11 go requirement?

> can you use something like GVM ? I usually install multiple go versions.

everytime i try to install something, something breaks, but i can try.
(linux is fun)

> It looks like you're passing verify now anyhow though

yes, verify does not pass locally now.

File: pkg/cluster/kubeadm/config.go [hunk 5]
```diff
@@ -51,25 +52,35 @@ func (c *ConfigData) Derive() {
...
+// API version v1alpha3
+const DefaultConfigTemplateAlphaV3 = `# config generated by kind
+apiVersion: kubeadm.k8s.io/v1alpha3
+kind: ClusterConfiguration
 kubernetesVersion: {{.KubernetesVersion}}
 # TODO(bentheelder): fix this upstream!
-# we need nsswitch.conf so we use /etc/hosts
-# https://github.com/kubernetes/kubernetes/issues/69195
-apiServerExtraVolumes:
-- name: nsswitch
```
Comment (BenTheElder) [inline pkg/cluster/kubeadm/config.go]: can you re-introduce the nsswitch bit? I'll follow-up with figuiring out how to remove it in the future, but right now it's required to make cluster up work for k8s master branch

Comment (neolit123) [reply pkg/cluster/kubeadm/config.go]: missed deleting that too, my bad.

**Comment by BenTheElder** [comment]
One issue with nsswitch, otherwise looks great. Thanks!!

**Comment by BenTheElder** [comment]
(the pr log links are broken for the e2e, just pushed a fix to that to test-infra, next round of testing should handle it correctly)

**Comment by neolit123** [comment]
updated again. waiting on CI.

as a side note, `unifiedControlPlaneImage` might be removed in the next API iteration.

**Comment by BenTheElder** [comment]
Let's just remove that then 👍

**Comment by neolit123** [comment]
> Let's just remove that then +1

in this PR?

**Comment by BenTheElder** [comment]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/pull/sigs.k8s.io_kind/46/pull-kind-conformance-parallel/8 cluster is up 🎉

**Comment by BenTheElder** [comment]
This PR or a follow up. I'd like to get this in asap to unblock CI signal

**Comment by neolit123** [comment]
> This PR or a follow up. I'd like to get this in asap to unblock CI signal

added a commit here.

**Pull request merged**